src/Features/Core/Portable/SplitOrMergeIfStatements/Consecutive/AbstractMergeConsecutiveIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,197 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractMergeConsecutiveIfStatementsCodeRefactoringProvider
+        : AbstractMergeIfStatementsCodeRefactoringProvider
+    {
+        // Converts:
+        //    if (a)
+        //        Console.WriteLine();
+        //    else if (b)
+        //        Console.WriteLine();
+        //
+        // To:
+        //    if (a || b)
+        //        Console.WriteLine();
+
+        // Converts:
+        //    if (a)
+        //        return;
+        //    if (b)
+        //        return;
+        //
+        // To:
+        //    if (a || b)
+        //        return;
+
+        protected sealed override CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument, string ifKeywordText)
+            => new MyCodeAction(createChangedDocument, ifKeywordText);
+
+        protected sealed override async Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifLikeStatement, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var ifGenerator = document.GetLanguageService<IIfLikeStatementGenerator>();
+
+            return CanBeMergedWithParent(syntaxFacts, ifGenerator, ifLikeStatement) ||
+                   await CanBeMergedWithPreviousStatementAsync(document, syntaxFacts, ifGenerator, ifLikeStatement, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected sealed override SyntaxNode GetChangedRoot(Document document, SyntaxNode root, SyntaxNode ifLikeStatement)
+        {
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            var ifGenerator = document.GetLanguageService<IIfLikeStatementGenerator>();
+            var generator = document.GetLanguageService<SyntaxGenerator>();
+
+            var isElseIfClause = ifGenerator.IsElseIfClause(ifLikeStatement, out var parentIfLikeStatement);
+            var previousIfLikeStatement = isElseIfClause ? parentIfLikeStatement : GetPreviousStatement(syntaxFacts, ifLikeStatement);
+
+            var newCondition = generator.LogicalOrExpression(
+                ifGenerator.GetCondition(previousIfLikeStatement),
+                ifGenerator.GetCondition(ifLikeStatement));
+
+            newCondition = newCondition.WithAdditionalAnnotations(Formatter.Annotation);
+
+            root = root.TrackNodes(previousIfLikeStatement, ifLikeStatement);
+            root = root.ReplaceNode(
+                root.GetCurrentNode(previousIfLikeStatement),
+                ifGenerator.WithCondition(root.GetCurrentNode(previousIfLikeStatement), newCondition));
```

Comment:
i feel like i'm not understand the need to:
1. track nodes.
2. replace the root.
3. then use an editor.

Can you clarify why all this is needed (ideally with comments)? Thanks!

Reply:
I may not have understood the need for this either. I think we need to use the editor on a new root because if we did all this in the editor including changing the if statement condition here, we'd then be unable to do the next steps like replacing and removing. They have to be done in order.

Reply:
Comments added in 7325a160fb3350ed422862950d34e39701e0bdd0